pkg/control/sidecarcontrol/util.go

```diff
@@ -270,13 +274,80 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
 	}
 
 	for _, tEnv := range sidecarContainer.TransferEnv {
-		key := fmt.Sprintf("%v/%v", tEnv.SourceContainerName, tEnv.EnvName)
-		env, ok := envsInPod[key]
-		if !ok {
-			klog.Warningf("there is no env %v in container %v", tEnv.EnvName, tEnv.SourceContainerName)
-			continue
+		envs := sets.NewString()
+		if tEnv.EnvName != "" {
+			envs.Insert(tEnv.EnvName)
+		}
+		for _, e := range tEnv.EnvNames {
+			envs.Insert(e)
+		}
+
+		sourceContainerName := tEnv.SourceContainerName
+		if tEnv.SourceContainerNameFrom != nil && tEnv.SourceContainerNameFrom.FieldRef != nil {
+			containerName, err := ExtractContainerNameFromFieldPath(tEnv.SourceContainerNameFrom.FieldRef, pod)
+			if err != nil {
+				klog.Errorf("unmarshal pod(%s.%s) annotations[%s] failed: %s", pod.Namespace, pod.Name, err.Error())
+				continue
+			}
+			sourceContainerName = containerName
+		}
+		// if sourceContainerName is empty or not found in pod.spec.containers
```

Comment:
"if sourceContainerName is empty or not found in pod.spec.containers" is this comments valid here, consider move above the warning log